### ISSUE TITLE
Tracking Issue for inherent unchecked integer methods

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust#85122]
This is a tracking issue for the `unchecked_*` methods on integers.

## `unchecked_math` (stable as of #122520)

```rust
impl uN {
    pub const unsafe fn unchecked_add(self, rhs: uN) -> uN;
    pub const unsafe fn unchecked_sub(self, rhs: uN) -> uN;
    pub const unsafe fn unchecked_mul(self, rhs: uN) -> uN;
}
impl iN {
    pub const unsafe fn unchecked_add(self, rhs: iN) -> iN;
    pub const unsafe fn unchecked_sub(self, rhs: iN) -> iN;
    pub const unsafe fn unchecked_mul(self, rhs: iN) -> iN;
}
```

### Steps / History

- [x] #69659 added `unchecked_{add,sub,mul}`
- [x] #85096
- [x] #115626
- [x] #122520
- [x] #121571

## `unchecked_neg`

```rust
impl iN {
    pub const unsafe fn unchecked_neg(self) -> iN;
}

```

### Steps / History

- [x] #115626 added `unchecked_neg`
- [x] #121571

## `unchecked_shifts`

```rust
impl uN {
    pub const unsafe fn unchecked_shl(self, rhs: u32) -> uN;
    pub const unsafe fn unchecked_shr(self, rhs: u32) -> uN;
}
impl iN {
    pub const unsafe fn unchecked_shl(self, rhs: u32) -> iN;
    pub const unsafe fn unchecked_shr(self, rhs: u32) -> iN;
}
```

### Steps / History

- [x] #85703
- [x] #103456
- [x] #115626
- [x] #121571

## `unchecked_*` (aggregate)

```rust
impl uN {
    pub const unsafe fn unchecked_add(self, rhs: uN) -> uN;
    pub const unsafe fn unchecked_sub(self, rhs: uN) -> uN;
    pub const unsafe fn unchecked_mul(self, rhs: uN) -> uN;
    pub const unsafe fn unchecked_shl(self, rhs: u32) -> uN;
    pub const unsafe fn unchecked_shr(self, rhs: u32) -> uN;
}
impl iN {
    pub const unsafe fn unchecked_add(self, rhs: iN) -> iN;
    pub const unsafe fn unchecked_sub(self, rhs: iN) -> iN;
    pub const unsafe fn unchecked_mul(self, rhs: iN) -> iN;
    pub const unsafe fn unchecked_shl(self, rhs: u32) -> iN;
    pub const unsafe fn unchecked_shr(self, rhs: u32) -> iN;
    pub const unsafe fn unchecked_neg(self) -> iN;
}
```

### Steps / History

- [x] #69659 added `unchecked_{add,sub,mul}`
- [x] #85096
- [x] #85703
- [x] #103456
- [x] #115626 added `unchecked_neg`
- [x] #122520
- [x] #121571
- [ ] Final commenting period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Should the other unchecked intrinsics like `exact_div` also get inherent versions?
- From #85703, consider if distinguishing the different kinds of ub could be useful (UB from overflow like `MIN/-1` or `MAX+1`, LLVM's `n[us]w`; UB from input range like `x/0` or `x << -1`; UB from lossy like `2/4` or `3 >> 1`, LLVM's `exact`)

Resolved unresolved questions:
- <del>Is `unchecked_*` the best naming for these?</del> <ins>We stabilised `unchecked_{add,sub,mul}` already, so, yes.</ins>

**Comment by jordens** [comment]
Thanks for this! We are seeing some noticeably faster code in on several embedded DSP applications in cortex-m7 with `unchecked_sh[lr]`. The additional masking has always bothered me as unnecessary in our case.

**Comment by jordens** [comment]
One question though:

#85703 adds `unchecked_sh[rl](self, rhs: Self)` while all the other take a `u32`: `{overflowing,wrapping,checked}_sh[rl](self, rhs: u32)`. Why is that?

**Comment by cmpute** [comment]
I vote for adding `exact_div` into the library! It could be heavily used in integer factorization and other applications. I'm also developing a checked version of exact division in my crate. I think `unchecked_exact_div` or `unchecked_div_exact` will be good names.

Besides, I'm also thinking about a shorter name for `unchecked_`, something like `raw_add`, `raw_shl` may be easier to type?

**Comment by clarfonthey** [comment]
I don't think that "raw" is a good naming scheme, since in general, you should *not* be using these primitives directly. In general, they should be hidden behind other methods that are safe to call. So, I would say that the "unchecked" naming is, at best, not descriptive enough, and that "raw" would be further from the goal of being descriptive.

**Comment by scottmcm** [comment]
@jordens Agreed the different RHS type is incorrect.  I've sent PR #103456.

**Comment by jdahlstrom** [comment]
I was looking for an `unchecked_div` the other day and was surprised that it doesn't exist even under this feature. Any particular reason it's not included?

**Comment by scottmcm** [comment]
I think that if there's a reason, it's that there two kinds of unchecked-ness in signed unchecked_div, and it'd be very easy to trigger the `INT_MIN / -1` UB when you only remembered to prove that the denominator is non-zero.

I definitely prefer an unsafe type invariant to an unsafe operation where possible, so I think I like the `Div<NonZeroTTT>` better.  After all, with #100428 I think it's `x / NonZero::new_unchecked(y)`, which is very competitive with `x.unchecked_div(y)`, especially if the producer of `y` can just change to reflect the non-zeroness of its result.

But it might fine to add `unchecked_div` under this gate and leave it a stabilization question to figure it out if it's worth keeping.  We'll see what libs-api has to say.

(Notably, the "the addition overflows" kind of UB is the only kind of UB for `unchecked_add`, and it's also something that can't be feasibly pre-checked and thus put into a type invariant, the way "it's not zero" can be.)

**Comment by ojeda** [comment]
For the signed cases, I think it is good to have it for consistency with the other operations and to avoid having to play with `unreachable_unchecked()` (and hope it gets optimized away) or `intrinsics` (unstable). Or is there another way to do it in stable?

For the unsigned cases, I think it would still be good to offer it for consistency with the signed case and with the other operations, even if one can use `NonZero*` (perhaps documenting that as an alternative).

**Comment by RalfJung** [comment]
Are there any objections to stabilizing at least `unchecked_{add,sub,mul}`? For those there shouldn't be any surprised about what their safety requirements are.

**Comment by clarfonthey** [comment]
A unary `unchecked_neg` could probably also be added to that list.

**Comment by scottmcm** [comment]
@RalfJung That seems entirely reasonable to me.  `a.unchecked_add(b)` is semantically `a.checked_add(b).unwrap_or_else(hint::unreachable_unchecked)`, so I don't think there's any lang issues with it -- it's not a newly expressible concept, not even in `const` -- so I think a libs-api FCP and it'd be good to go.

Nominating it on behalf of Ralf to get libs-api's opinion.  Would you want the FCP the whole thing, or FCP a PR that just does a subset?

**Comment by m-ou-se** [comment]
We discussed this in a recent libs-api meeting. We mostly felt positive about stabilizing this, but it'd be helpful to have a clear overview of the exact APIs this tracking issue is tracking.

**Comment by RalfJung** [comment]
Great; does someone want to prepare a stabilization PR? I think that can cover everything that is currently under the `unchecked_math` feature gate; an `unchecked_div` function does not seem to exist (it only exists as an intrinsic, and we also have the `exact_div` intrinsic, but those will obviously not get stabilized).

> A unary `unchecked_neg` could probably also be added to that list.

Such a function doesn't exist yet, so it would have to be added (in a separate PR) before it can be stabilized. This requires an ACP since it's a new public function. None of this is insurmountable but I think if someone wants to push for this they can do this in parallel to the effort of stabilizing the existing functions in the `unchecked_math` feature gate; there's no reason they have to all be stabilized at once.

**Comment by clarfonthey** [comment]
I was under the impression that since we have a tracking issue for this, adding a related function (like `unchecked_neg`) would be part of the general API ironing-out process and thus not require an ACP. Effectively, I'd expect the response to the ACP to be "this discussion should go in the tracking issue," so, that's why I'm discussing this here.

If we were to propose radically different APIs it would make sense to make a new ACP, but since we're discussing the basic arithmetic operations that have already been marinating for a while, it makes sense to simply make a PR to add `unchecked_neg`.

**Comment by RalfJung** [comment]
:shrug: I'm not on the libs-api team, feel free to give it a shot. :)

**Comment by unia909** [comment]
Hello. I found the problem in using unchecked integer methods. My program works correctly in debug mode and outputs INT3 at the end, but in release mode the program skips my check by 0, causing the program to not work correctly. If I replace `unchecked_add` with `wrapping_add`, the program works correctly in both release and debug modes. This problem is also repeated on rust playground.
```rust
#![feature(unchecked_math)]

use std::convert::TryInto;

#[derive(Debug)]
pub struct CPU {
    ram: Vec<u8>
}

impl CPU {
    pub fn new(ram: Vec<u8>) -> Self {
        Self { ram }
    }

    pub fn exec(&mut self, mut ip: usize) {
        let mut di = 0u16;
        let mut zero = false;
        loop {
            let current = self.ram[ip];
            ip += 1;
            match current {
                0x47 => {
                    println!("INC DI");
                    di = unsafe { di.unchecked_add(1) };
                    if di == 0 {
                        zero = true;
                    } else {
                        zero = false;
                    }
                }
                0x75 => {
                    let address: i8 = unsafe { core::mem::transmute(self.ram[ip]) };
                    let address = address as isize;
                    ip += 1;
                    println!("JNZ ({}) {address}", !zero);
                    if !zero {
                        if address < 0 {
                            ip -= (-address) as usize;
                        } else {
                            ip += address as usize;
                        }
                    }
                }
                0xBF => {
                    let val = u16::from_le_bytes(self.ram[ip..ip+2].try_into().unwrap());
                    ip += 2;
                    println!("MOV DI, {val}");
                    di = val;
                }
                0xCC => {
                    println!("INT3");
                    return;
                }
                0xC6 => {
                    match self.ram[ip] {
                        5 => {
                            let val = self.ram[ip];
                            ip += 1;
                            println!("MOV BYTE [{}], {val}", di);
                            let addr = di as usize;
                            self.ram[addr] = val;
                        }
                        _ => todo!()
                    }
                    ip += 1;
                }
                _ => todo!()
            }
        }
    }
}

fn main() {
    let mut ram = vec![0; 64 * 1024];
    /*
    mov di, 8000h
    loop:
    mov byte [di], 13h
    inc di
    jnz loop
    int3
    */
    ram[0..10].copy_from_slice(&[0xbf, 0x00, 0x80, 0xc6, 0x05, 0x13, 0x47, 0x75, 0xfa, 0xcc]);
    CPU::new(ram).exec(0);
}
```

**Comment by RalfJung** [comment]
@unia909 thanks for the report! However, this is a tracking issue, which are for discussing the feature in general, not any particular issues. Please open a new issue for your problem, and link it from here.

**Comment by RalfJung** [comment]
That said, there's a bug in your code. If you replace `di.unchecked_add(1)` by `di.checked_add(1).unwrap()`, you get an assertion failure. Using `unchecked_add` you are promising that this will never overflow, and yet it does overflow, so your code has UB and getting the wrong result is expected.

**Comment by unia909** [comment]
If I want the code to be able to overflow a number, then what should I use? Wrapping methods is the only correct way?

**Comment by RalfJung** [comment]
Yes, that's *exactly* what `wrapping_add` is for. Please carefully read the docs for `unchecked_add`. They explicitly say that you must not call `unchecked_add` when things can overflow!

**Comment by clarfonthey** [comment]
The PR fixing the feature flags is merged and we could probably do an FCP for the `unchecked_math` methods, assuming that the recent addition of `unchecked_neg` isn't too controversial. It's only implemented for signed integers because I didn't think that an "eat my laundry if this integer isn't zero" operator was a very good addition.

**Comment by comex** [comment]
I seem to recall the naming of these methods was bikeshedded long ago, but this does seem like a good example of the naming being confusing.

After all, the semantics of the normal + operator are 'wrapping addition plus an optional overflow check'.  It's easy to guess that a method called `unchecked_add` would have the semantics 'same as + but without the check'.  (Compare to the similarly named command-line option `-C overflow-checks=off`, which *does* have that effect.)

There's nothing wrong with the name of the correct method, `wrapping_add`.  But some users may not be familiar with the term 'wrapping'.  Or even if they do know what it means, they may be tunnel-visioned on the term 'overflow' (since that's what the error said when they tried to use regular addition).  Or they may just not notice `wrapping_add` in the method list.

If nothing else, the documentation could be improved:
- The documentation for `unchecked` methods should have a more explicit caution and should point you to the `wrapped` methods.  

   After all, they should be used extremely rarely in idiomatic Rust.  And they may be confusing since there are almost no other programming languages with 'UB on overflow' operations – with the obvious exception of C and C++, but in their case only for signed integers.  (I do notice that unia909's example was using an unsigned integer.  In the abstract it might seem obvious that signedness of integers is orthogonal to wrappingness of arithmetic, but it may not feel that way if you're used to C or C++.)

- The documentation for `wrapping` methods should perhaps mention the term 'overflow', especially to assist anyone doing a find-in-page on the docs site.

- On a similar note, users who search for 'overflow' are even more likely to find the `overflowing` methods, so the docs for those should perhaps explicitly mention `wrapping` as an alternative to use if you don't want the boolean.

**Comment by clarfonthey** [comment]
I honestly don't know how the docs could be more clear than:

> Unchecked integer addition. Computes `self + rhs`, assuming overflow cannot occur.

We could also link to the `wrapping` methods in the docs, but ultimately, I don't think that it's worth renaming the methods to something non-standard (we already have many similar "unchecked" methods that have invariants in the docs) since the real problem here is people using unsafe code without knowing what they're doing.

**Comment by clarfonthey** [comment]
One thing that hadn't occurred to me that might improve the situation is using the `assume_unsafe_precondition!` macro inside these methods, like is done in similar libstd methods. Will submit a PR to see if that improves things.

**Comment by ojeda** [comment]
One way to alleviate the concern of misuse would be to have a feature in `rustdoc` to toggle a button to show "expert"/"uncommon" methods. For already `unsafe` methods, it may not be that needed, but this could be quite useful for other "safe but subtle" methods. This could be used not just for things related to UB, but also other kinds of methods that may be tricky for other reasons. It could be generalized further (e.g. categories or tags) that can be useful in libraries (e.g. low-level vs. high-level API).

Personally, I really like that Rust has this kind of unsafe methods and provides choice -- it is something I have mentioned to the C committee, as an example of what it could be provided.

**Comment by RalfJung** [comment]
> One thing that hadn't occurred to me that might improve the situation is using the `assume_unsafe_precondition!` macro inside these methods, like is done in similar libstd methods. Will submit a PR to see if that improves things.

I agree it's good to have these, but they wouldn't have helped here since the std we ship to users is built without debug assertions.

**Comment by clarfonthey** [comment]
> > One thing that hadn't occurred to me that might improve the situation is using the `assume_unsafe_precondition!` macro inside these methods, like is done in similar libstd methods. Will submit a PR to see if that improves things.
> 
> I agree it's good to have these, but they wouldn't have helped here since the std we ship to users is built without debug assertions.

I was under the impression that this macro was specifically meant to be inlined so that it *does* apply without debug assertions, although I am probably wrong about that. No idea how `cfg!` interacts with inlining.

> One way to alleviate the concern of misuse would be to have a feature in `rustdoc` to toggle a button to show "expert"/"uncommon" methods. For already `unsafe` methods, it may not be that needed, but this could be quite useful for other "safe but subtle" methods. This could be used not just for things related to UB, but also other kinds of methods that may be tricky for other reasons. It could be generalized further (e.g. categories or tags) that can be useful in libraries (e.g. low-level vs. high-level API).

I have no data for this, but I would assume that the people who are calling this method in error are doing so primarily via IDE completions, guessing the method, or misguided copy-paste, and thus this wouldn't really help. We could add a warning to the compiler that triggers whenever a user uses an unsafe block for the first time and require them to agree to an "I pinky-promise I can read the docs and will" message, but that just feels patronising and like a lot of work for little gain.

**Comment by RalfJung** [comment]
> I have no data for this, but I would assume that the people who are calling this method in error are doing so primarily via IDE completions,

Sounds like we should work with IDE people to avoid auto-completing unsafe functions outside of unsafe blocks, or at least somehow do that in a way that makes it clear to users that they need to take manual action here.

**Comment by clarfonthey** [comment]
The main reason is that the other arithmetic methods use prefixes (e.g. `wrapping_add`), although, fair point.

It's another weird inconsistency we'll probably have to live with.

**Comment by Kixunil** [comment]
As was mentioned, I agree that the name is confusing for the exact reason @comex explained. I think I've seen an experienced developer getting confused about the meaning (not sure if just communication issue).

Proposed alternatives:

* `add_assume_no_overflow`
* `add_no_overflow`
* `non_overflowing_add`

They are bit longer but I guess the method is not used a lot.

**Comment by workingjubilee** [comment]
`inbounds_{add,mul,sub}`.

**Comment by clarfonthey** [comment]
I mean, it seems pretty fine to have `unchecked_*` methods that are just the `checked_*` methods except that the `None` case is UB. IMHO, the confusing bits are where these don't exactly overlap with the way the `checked_*` methods work, like for division where you're not sure if you're talking about divide-by-zero, overflow, or a remainder.

**Comment by workingjubilee** [comment]
I personally don't _mind_ the `unchecked_*` names, but I do love to go through my Pantone book!

**Comment by clarfonthey** [comment]
Note that `unchecked_div` is also not under this feature flag, if it does exist. (I don't think it does beyond the intrinsic?)

I'd be fine with making the documentation update before stabilisation, or as part of it, though. I think it's worth pointing out.

**Comment by ojeda** [comment]
> > One thing that hadn't occurred to me that might improve the situation is using the `assume_unsafe_precondition!` macro inside these methods, like is done in similar libstd methods. Will submit a PR to see if that improves things.
> 
> I agree it's good to have these, but they wouldn't have helped here since the std we ship to users is built without debug assertions.

With https://github.com/rust-lang/rust/pull/120594 merged now everyone would benefit so...

> Could we document every single `a.unchecked_foo(b)` as being identical to `a.checked_foo(b).unwrap_unchecked()`?
>
> Are there any exceptions to that rule?

...`-Cdebug-assertions=y` makes them differ now (for everybody, not just users that build `core`), even though I don't know if those checks are documented as a guarantee (e.g. `unwrap_unchecked()` did not document its former `debug_assert!`): https://godbolt.org/z/W4ob6eorE.

Should `unchecked_add` etc. get the unsafe debug checks too? Cc @saethlin. I imagine that it would make sense, since even `unreachable_unchecked` has it now.

But then how does one write a "unchecked even in debug" one? That is useful in projects where debug builds require some performance to be useful, and certain checks may need to be disabled for that reason. `checked_foo()` + `match` + intrinsics works in this instance, but it requires intrinsics. And what if there is no "checked" version of an API?

Is there another way I am missing?

**Comment by clarfonthey** [comment]
The unsafe preconditions are designed to be as slim as possible for performance reasons, so, hopefully shouldn't make things too slow, but I also am not sure either way.

**Comment by ojeda** [comment]
They will make things too slow for somebody sooner or later -- it is not uncommon in projects to have a "fast debug" build where a few checks need to be explicitly bypassed in hot paths. And it is not just the branch/call itself, it can also break other optimizations like vectorization: https://godbolt.org/z/MMTordabn.

I really like the unsafe precondition checks, but I would like to make sure there is an escape hatch for `-Cdebug-assertions=y` builds. Perhaps we need a `core::hint::unreachable_unchecked_even_in_debug`-like (even if it does not solve the "there is no checked API" case, but I don't know if that is an issue). For me, `core::hint::unreachable_unchecked` was that in the past, and things like `unwrap_unchecked()` were the "checked in debug" ones.

I also just noticed that using the intrinsic for slice indexing is not enough to remove the check in a debug build and get vectorization back (included in the link).

(I guess it is best if I move this into its own issue).

**Comment by RalfJung** [comment]
> (I guess it is best if I move this into its own issue).

Indeed, this is largely unrelated to this issue and mostly about https://github.com/rust-lang/rust/pull/120594, so please open a new issue.

**Comment by saethlin** [comment]
I started typing out a long reply, but I'll wait for a new issue. For now, please check https://github.com/rust-lang/rust/issues/120848, I've already thought of quite a few follow-up tasks on these checks and wouldn't mind adding to the list.

**Comment by clarfonthey** [comment]
As an aside to specifically mention the status of adding unsafe checks to these methods, #117494 was stalled because of compiler errors I couldn't resolve, which lead to me focusing on #117571 to attempt to provide more info to said errors so I could debug them.

I haven't done any work on these PRs recently, but I'll try to rectify that this week, and if anyone else is interested in taking a look to maybe help out, I would appreciate it. I'm very unfamiliar with compiler internals, so, it's been a bit of a learning curve for me.

**Comment by clarfonthey** [comment]
More up-to-date update: I decided to get around the compiler lint errors for now, and while it did require a bit of excursion for the shift operations in particular, #117494 should be ready to merge and will panic in debug mode when the unsafe assertions are ignored. That should hopefully help substantially with the concerns that people are using these methods incorrectly, since users without any additional setup to disable these assertions (who should not be doing that unless they know what they're doing) will run into errors in their tests when using these incorrectly.

**Comment by the8472** [comment]
Unchecked variants of the `{signed_integer}::checked_{add,sub}_unsigned` functions and vice-versa could be useful if those properties can be conveyed to LLVM.

See #122420

**Comment by clarfonthey** [comment]
Cleaned up this issue a bit, separating out the various feature flags better and including more PRs that were omitted from the list. #121571 should hopefully be merged soon to ensure that all of these have unsafe preconditions checked in debug mode.

**Comment by clarfonthey** [comment]
Worth an extra update: the unsafe preconditions PR is finally merged! Won't ping everyone who helped get this over the line here since there were quite a few, but this means that in the next nightly all unchecked arithmetic using the methods (not the intrinsics) should be checked in debug mode.

**Comment by rust-rfcbot** [comment]
:bell: **This is now entering its final comment period**, as per the [review above](https://github.com/rust-lang/rust/issues/85122#issuecomment-3383570036). :bell:

**Comment by rust-rfcbot** [comment]
The final comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/issues/85122#issuecomment-3383570036), is now **complete**.

As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

This will be merged soon.

**Comment by cuviper** [comment]
AFAICS #149087 was the last of this.
